README.md
# javascript-synchronized-clock
A very accurate clock for display on web pages.

https://kwynn.com/t/9/12/sync/ - live version

It's JavaScript on the front end and PHP on the back end.  The JS polls the 
server (PHP) for time.  

There is a much more detailed readme.html in the repo root.

2020/12 recap - This does not work in Firefox mobile or perhaps any Safari.
It works with Chrome mobile.

I made a mistake in creating class variables that I didn't need.  I just 
want object variables, not static class variables.  Perhaps one day I'll fix this.

Several major mobile browsers do not support class variables.


CHANGELOG.md
2020/12/30 1:37am EST

I got rid of the local kwutils.php
That file is now part of https://github.com/kwynncom/kwynn-php-general-utils

I also got rid of references to that file.  I also had to get rid of 
references to /sys/hypervisor/uuid  That is no longer any help in 
determing AWS.

Amazingly the crypto AWS test now works again both on the web and CLI.  Amazing 
that the fix was that simple.

This must be almost exactly a year later.  I was somewhat surprised to remember 
that kwutils.php started with this project.  


js/kwClockSynced.js
if (typeof byid !== 'function') { function byid(id) { return document.getElementById(id); } }

var KWCLSY_TESTMODE = 0;

class kwclsyOntick { // execution in this file starts here - this is the controller class
    
    doticko = 0;    // dotick object
    synccnt = 15;   // sync count (number of time polls to make to the time server)
    theoff = false; // the time offset or the correction or error between client and server
    dispo;	    // display / output object
    networko;	    // network object -- actually does the poll
    icorro;	    // internal error object -- accounts for slower cell phones getting behind the clock tick
     
    constructor() {
	this.dispo = new kwclsyDetailedDisplay();
	
	this.dispo.initTick();
	
	this.icorro = new kwclsyInternalCorrectionO();	     // internal error tracking object
	this.dispo.setCBO(this);		     // give the display this callback object for resync purposes
	this.doticko = new kwclsyDoTickC(this); // the clock tick itself object / class
	this.networko = new kwclsyNet(this.synccnt, this, this.dispo); // object that does polling
	this.networko.doSync(); // start polling the server
    }
    
    resync(reset)    { this.networko.doSync(reset); }
    doOnSync(offset) { this.theoff = offset; }
    
    // corrected time, local (uncorrected) time, from-tick internal error.  I cycle the internal error around because the next correction is based on 
    // the previous correction
    doOnTick(corrTime, localTime, ticorr) {
	let   ats  = this.dispo.ditick(corrTime, localTime, ticorr, this.theoff); // display a tick; ats is after-display timestamp
	// ats accounts for processing between the calculated clock tick and actual display, or internal correction (icorr)
	
	const next = this.doticko.doTick(this.theoff, this.icorro.get());  // calc next tick, which invokes this function again
	this.dispo.prep(next); // cache the next tick's string / output
	
	if (KWCLSY_TESTMODE) ats = this.dispo.getATS();
	
	this.icorro.put(corrTime, ats, ticorr, this.theoff);
    }
}

class kwclsyDoTickC { // calculates and does the next tick
    pinterval = 1000; // permanent interval or "tick" period; 1000 ms = 1s.  Note that JS does time natively in ms
    mini = 250;       // under this calculated interval, skip to the next second
    onto;             // on tick object

    constructor(o) { 
	if (!o) return; // I also invoke this from the disp object just to get the interval
	this.onto = o;	 
	this.doTick(false, 0); // do first tick with no sync data yet
    }
    
    // calculates and executes the setTimeout interval until the "top of the second" -- n seconds and 0 ms
    // on slower phones, the difference between the "tick" and the display of the tick is long enough to correct for an internal error
    doTick(ecorr, icorr) {  // external and internal correction
	
	const lnow = new Date().getTime(); // local now / time
	let cnow = false;                  // corrected time
        if (ecorr !== false) cnow = lnow + ecorr; 
	const permi = this.pinterval; // This makes the following a bit more readable
	let cmpnow = lnow;            // comparistaton "now" -- which time to use for tick
	if (cnow) cmpnow = cnow;
	let dint = permi - (cmpnow % permi); // the displayed interval
	if (dint < this.mini) dint += permi;
	
	let aint = dint - icorr;	    // fire the next tick corrected for internal delay / error

	let cnowDi = false;		    // corrected time to be displayed
	if (cnow !== false) cnowDi = cnow + dint; // this is what will be displayed after the tick executes--the next second
	
	const self = this;
	setTimeout(function() {	
	    // similarly the next local (uncorrected) time is lnow + aint
	    self.onto.doOnTick(cnowDi, lnow + aint, icorr); 
	}, aint);
	
	return cnowDi; // as a matter of cacheing, return the next second to be displayed
    }
    
    getInterval() { return this.pinterval; }
}

class kwclsyNet { // does the network operations -- polling and gets chrony info
    
    disp = false; // display object
    estato;	  // external (server) statistics object
    sycmax;	  // number of polls to do
    cbo;	  // callback object (the "controller")

    syc = 0;         // sync count - polls done
    sysetc = 0;      // number of sets of syc done
    
    constructor(sycin, cboin, disp) {
	this.disp = disp;
	this.sycmax = sycin;
	this.cbo = cboin;
	this.estato =  new kwclsyStats();
    }
    
    doSync(reset) {
	this.getChronyInfo(reset);
	this.poll();
    }
    
    getChronyInfo(reset) {
	
	if (reset) this.doReset();
	
	const xhr = new XMLHttpRequest(); 
	xhr.open('GET', 'services/getChrony.php');
	let self = this;
	xhr.onreadystatechange = function() {
	    if (this.readyState !== 4) return;
	    let chronyInfo = false;
	    if (this.status === 200) chronyInfo = JSON.parse(xhr.response);
	    self.disp.showChronyInfo(chronyInfo); // display chrony info
	}
	
	xhr.send();
    }
    
    poll() {
	const xhr = new XMLHttpRequest(); 
	xhr.open('GET', 'services/getTimeSimple.php');
	const self = this;
	xhr.onreadystatechange = function() {
	    if (this.readyState !== 4 || this.status !== 200) return;
	    const tsa = new Date().getTime(); // timestamp after receipt
	    const tss = self.srvTimeToCliTime(xhr.response); // timestamp server
	    const thecorr = ((tss - tsb) + (tss - tsa )) / 2; // a version of the SNTP offset calculation
	    self.estato.put(thecorr, tss);
	    self.disp.showStatInfo(self.estato.getAll()); // display stats
	    
	    // keep polling until you hit the set max number of polls
	    if (++self.syc < (self.sycmax * (self.sysetc + 1)) ) self.poll();
	    else {
		self.sysetc++;
		self.cbo.doOnSync(self.estato.getOff(), self.disp); // now we're ready to display a corrected clock
	    }
	}
	const tsb = new Date().getTime(); // timestamp before poll
	xhr.send();
    }

    doReset() {
	this.syc = this.sysetc = 0;
	this.estato.reset();
    }
    
    srvTimeToCliTime(st) { // server emits 1579551951.9446 s which becomes 1579551951945 ms
	const  ct = Math.round(st * 1000);
	return ct;
    }    
}

class kwclsyStats {
    
    aaa = []; // average array.  I made it aaa for distinctiveness
    aai = 0;  // iterator
    theavg = 0; // the average offset
    sd = 9999;  // standard deviation, start with something "absurd."
    latestTS = 0; // timestamp keeping track of how old the info is

    reset() {
	this.aai = 0;
	this.aaa = [];
	this.theavg = 0;
	this.sd = 9999;
	this.latestTS = 0;	
    }    

    put(xi, ts, maxi) { // x subscript i for the standard dev calc

	this.latestTS = ts;
	
	const aai = this.aai++;
	if (maxi) this.aaa[aai % maxi] = xi; // rotate the "buffer" / limit how much we're keeping
	else      this.aaa[aai]        = xi; 
	let sum = 0;
	
	let aaal = this.aaa.length;
	for (let i=0; i < aaal; i++) sum += this.aaa[i];
	const avg = sum / aaal;
	this.theavg = avg;
	let sdsum = 0;
	for (let i=0; i < aaal; i++) sdsum += Math.pow(this.aaa[i] - avg, 2);
	
	if (this.aai >= 2) this.sd = Math.sqrt(sdsum /  (aaal - 1));
	else               this.sd = avg;

    }
    getOff()  { return Math.round(this.theavg); }
    getAll()  {
	const rob = {}; // return object
	rob.i   = this.aai;
	rob.avg = this.theavg;
	rob.sd  = this.sd;
	rob.ls  = this.latestTS;
	return rob;
    }
}

class kwclsyInternalCorrectionO {

    istato;    // internal statistics
    icorr = 0;
    
    constructor() { this.istato = new kwclsyStats(); }
    
    put(ct, af, pcorr, ecorr) { // corrected time, after-display time, previous correction, external (network) correction
	
	if (!af) return; // I believe it's only in test mode that this would be unset
	
	const max = 75;  // In very early versions I had bad experiences with trying to over-correct
	
	let newc = pcorr + ((af + ecorr ) - ct); // new correction
	if (newc < 0) newc = 0;
	if (newc > max) newc = max;
	
	this.istato.put(newc, af, 30); // limit to 30 data points; otherwise the correction would take too long
	this.icorr = this.istato.getOff();
    }
    
    get () { return this.icorr;  }
}

class kwclsyDetailedDisplay {

    ci1     = byid('ci1');
    avg     = byid('nativeError');
    sdd     = byid('sdd');
    spolls  = byid('spings');
    corrCl  = byid('kwTheClock');
    natCl   = byid('nativeClock');
    adCl    = byid('adClock');
    ciate   = byid('ciat');
    lsate   = byid('lsat');
    resybe  = byid('resyb');
    resybre = byid('resybr');
    accuse  = byid('accus');
    icorre   = byid('icorr');
    iicorre  = byid('iicorr');
    beatpe  = byid('beatPar');
    beates  = [];
    bmax    = 1;
    bdiv;
    lsat;
    ciat;
    status = 0;
    sycbo;
    ats = 0;
    cache = {};
    dipinterval;
    showMS  = false;
    lus = false;
        
    prep(nt) {
	if (!nt) return;
	this.cache[nt] = this.tick2(nt, false, 'corr');
    }
    
    constructor() {
	const ticko = new kwclsyDoTickC();
	this.dipinterval = ticko.getInterval();
	this.setResetHandlers();
	this.setBeats();
	this.prepDi();
    }
    
    setResetHandlers() {
	var self = this;
	const more = byid('resyb' );
	const rest = byid('resybr');
	if (more) more.onclick = function() { self.resync(0); }
	if (rest) rest.onclick = function() { self.resync(1); }
    }
            
    setCBO(o) { this.sycbo = o;  }
        
    showChronyInfo(cin) {
	
	// dif() is defined below -- "display if element exists"
	if (!cin || !cin.status || cin.status !== 'OK') { dif(this.ci1, 'Chrony info not received from server, so time reliability is uknown.'); return; }
	const basic = cin.basic;
	this.ciat = basic.asOfms;
	this.ciacc(basic.stms, basic.stus);
    }
    
    ciacc(ms, us) {
	
	let stdi;
	
	if (Math.abs(ms) < 1) stdi = '< 1';
	else stdi = roundTo(ms, 1);
	
	dif(this.ci1, stdi);
	dif(this.accuse, us);
	
    }
    
    showStatInfo(sin) {
	dif(this.avg, Math.round(sin.avg));
	dif(this.sdd, Math.round(sin.sd ));
	this.lsat = sin.ls;
	dif(this.spolls, sin.i);
    }
    
    getATS() { return this.ats; }
    
    ditick(corts, nowts, icorr, ecorr) {
	if (KWCLSY_TESTMODE) this.ditickt(corts, nowts, icorr, ecorr);
	else return   this.ditickr(corts, nowts, icorr, ecorr);
    }
    
    ditickt(corts, nowts, icorr, ecorr) {

	 if (this.ats) this.ats += this.dipinterval;

	    	const self = this;
	setTimeout(function () {
	   self.ditickr(corts, nowts, icorr, ecorr);    
	}, 10);
    }
    
    doLUS() { return this.corrCl && this.corrCl.dataset && this.corrCl.dataset.local_until_sync != false;  }
    
    ditickr(corts, nowts, icorr, ecorr) {
	    
	let chit = false;
	let ats = false;
	
	if (corts && this.cache[corts]) {
	    this.beat();
	    dif(this.corrCl, this.cache[corts]);
	    ats = new Date().getTime();
	    chit = true;
	}
	
	if (this.status === 0) {
	    if (this.resybe ) this.resybe .style.visibility = 'visible';
	    if (this.resybre) this.resybre.style.visibility = 'visible';
	    this.status = 1;
	}

	if      (corts && !chit)         ats = this.tick2(corts, this.corrCl, 'corr');
	else if (!corts && this.lus)           this.tick2(nowts, this.corrCl);
	
	if (1)           this.tick2(nowts, this.natCl);
	if (ats)         this.tick2(ats + ecorr, this.adCl);
		
	if (corts) {
	    if (this.ciat) 
	    dif(this.ciate, Math.round((corts - this.ciat) / 1000));
	    dif(this.lsate, Math.round((corts - this.lsat) / 1000));
	    dif(this.icorre, icorr);
	}
	
	this.ats = ats;
	
	if (corts && ats) dif(this.iicorre, ats + ecorr - corts);
	
	return ats;
    }
    
    initTick() { 
	if (this.lus) {
	    this.tick2(new Date().getTime(), this.corrCl);  } 
    }
    
    setBeatButtons() {
	const es = document.querySelectorAll('[data-kwclsy=beatChEle]');
	const self = this;
	es.forEach(function(e) {
	    e.onclick = function() { self.beatInp(this.value); }
	});
	
    }
    
    setBeats() {
	
	this.setBeatButtons();
	
	for (let i=0; i < 8; i++) this.beates[i]  = byid('beat' + i);
	this.setBI();
    }
    
    beat(iin) {
	
	if (this.bmax < 1) return;
	
	if (!iin) iin = 0;
	
	if (!this.beates || !this.beates[iin]) return;
	
	this.beates[iin].style.visibility = 'visible';
	const self = this;
	
	setTimeout(function() { 
	    self.beates[iin].style.visibility = 'hidden';
	    if (iin < self.bmax - 1) self.beat(iin + 1);
	}, self.bdiv);
    }
    
    beatInp(newb) {
	if (newb >= 1) {
	    this.beatpe.style.display = 'block';
	} else {
	    this.beatpe.style.display = 'none';
	}
	
	this.bmax = newb;
	
	this.setBI();
    }
    
    setBI() {
	if (this.bmax > 1) this.bdiv = this.dipinterval / this.bmax;
	else		   this.bdiv = 250;	
    }
    
    resync(reset) {
	this.resybe.style.visibility = 'hidden';	
	this.status = 0;
	this.sycbo.resync(reset);
	
    }
    
    prepDi() {
	if (this.corrCl &&
	    this.corrCl.dataset &&
	    this.corrCl.dataset.show_ms == true) this.showMS = true;
            // note that you need the   == true because '0' is considered true as a non-empty string
	    
	this.lus = this.doLUS();
    }
    
    tick2(tsin, ele, type) {
	
	if (ele && type === 'corr') this.beat();
	const dir = kwsynclDisplayTick(tsin, ele, type, this.showMS);
	if (ele && type === 'corr') {
	    const adts = new Date().getTime();
	    return adts;
	}
	else  if (type === 'corr')   return dir;
    }
} 

function roundTo(val, digits) {
    if (!digits) digits = 0;

    const pow = Math.pow(10, digits);
    const mul = val * pow;
    const rnd = Math.round(mul);
    const ret = rnd / pow;

    return ret;
}

if (typeof dif !== 'function') { // display if element exists
    function dif(ele, v) {
	if (!ele || (!ele.innerHTML && ele.innerHTML !== '')) return;
	ele.innerHTML = v;
    }
}


services/chrony.php
<?php // Kwynn Buess, kwynn.com, 2020/01/18 6:25pm my local

require_once('/opt/kwynn/kwutils.php');

class chrony_parse { // see "$ chronyc tracking" output and other doc at bottom

private static function getInternal() { 
    
    $res = shell_exec('chronyc tracking');
    
    // output will be an array with the left side of output as keys and right side as value
    
    $anl = explode("\n", $res); unset($res);
    $linec = count($anl);
    foreach($anl as $row) {
	$ac = explode(' : ', $row);
	if (!$ac || count($ac) !== 2) continue;
	if (   trim($ac[0]) &&  trim($ac[1]))
	    $a[trim($ac[0])] =  trim($ac[1]);
    }
    
    kwas(count($anl) - 1 === count($a) && count($a) === 13, 'array fail 1');
    
    return $a;
}
    
public static function get() {
    
    try {

    $a = self::getInternal();
    $ret = [];
    $ret['raw'] = $a;
        
    $r = $a['Reference ID'];
    preg_match('/([0-9A-Z]+) \(([^\)]+)\)/', $r, $matches);
    kwas(isset($matches[2]) && $matches[2] && is_string($matches[2]) && strlen(trim($matches[2])) > 5, 'rID regex fail');
    
    $sn = [];
    $sn['rid'] = $matches[2];
    
    if (isAWS() && $sn['rid'] === '169.254.169.123') $sn['rname'] = 'AWS EC2 Time Sync Service';
    else $sn['rname'] = '';
    
    $key = 'Ref time (UTC)';
    kwas(isset($a[$key]), 'no UTC ref time');
    
    $ts = strtotimeRecent($a[$key] . ' UTC');
    $sn['rts'] = $ts;
    
    $df = 'g:i:s A D m/j';
    $sn['rr']  = date($df, $ts); unset($ts);
    $sn['nr']  = date($df);
    
    $st = $a['System time'];
    
    preg_match('/(^\d+\.\d+) seconds (\w+) of NTP time/', $st, $matches); unset($st); kwas(isset($matches[2]), 'regex fail offset'); 
    
    $s = $matches[1];

    if      ($matches[2] === 'fast') $sign = '+';
    else if ($matches[2] === 'slow') $sign = '-';
    else kwas(0, 'not fast or slow');

    
    $fs = $sign . sprintf('%0.3f', $s * 1000000);
    
    $b = [];
    
    $b['stus'] = $fs;
    $b['stusr'] = round($fs);
    $b['std'] = $fs . '&#181;s';
    
    if ($sign === '-') $mult = -1;
    else               $mult =  1;
    
    $b['stms']   = $s   * 1000;
    $b['asOfms'] =         intval(round(microtime(1) * 1000, 0));
    $b['sts']    = ($mult >= 0 ? '+' : '-') . $s;
    $b['stats']  = time();

    $ret['basic'] = $b;
    $ret['status'] = 'OK';

    $key = 'Update interval';
    $sn[$key] = $a[$key];

    $ret['kwsntp'] = $sn;

    return $ret;
     
    }
    catch (Exception $ex) { 
	if (PHP_SAPI === 'cli') throw $ex;
	return false; 
    }
}
}
/* Regarding the Amazon (Web Services) Time Sync Service and 169.254.169.123, see
https://aws.amazon.com/blogs/aws/keeping-time-with-amazon-time-sync-service/

$ chronyc tracking
Reference ID    : A9FEA97B (169.254.169.123)
Stratum         : 4
Ref time (UTC)  : Fri Jan 17 03:22:28 2020
System time     : 0.000004590 seconds slow of NTP time
Last offset     : -0.000010681 seconds
RMS offset      : 0.000035949 seconds
Frequency       : 1.564 ppm slow
Residual freq   : -0.002 ppm
Skew            : 0.043 ppm
Root delay      : 0.000533058 seconds
Root dispersion : 0.001077776 seconds
Update interval : 1032.8 seconds
Leap status     : Normal
**************************
Note that I'm using chrony's names as short variable names above, such as rid for Reference ID and st for system time.

https://chrony.tuxfamily.org/

  */

js/kwClockGeneral.js
if (typeof dif !== 'function') {
    function dif(ele, v) {
	if (!ele || (!ele.innerHTML && ele.innerHTML !== '')) return;
	ele.innerHTML = v;
    }
}

function kwsynclDisplayTick(tsin, ele, type, showMS) { // see credit to Aaron Farrar in readme
   
    const displayDateO = new Date(tsin);
    let    h = displayDateO.getHours(); 
    let    m = displayDateO.getMinutes(); 
    let    s = displayDateO.getSeconds();
    const ms = displayDateO.getMilliseconds();
    if      (h === 0 ) h  = 12;
    else if (h   > 12) h -= 12;
    m = (m < 10) ? '0' + m : m;
    s = (s < 10) ? '0' + s : s;
    let st = h + ':' + m + ':';
    if (type === 'corr') st += '<span class="corrs">';
    st += s;
    if (type === 'corr') st += '</span>';
    if  (showMS)         st += '.' + (ms + '').padStart(3,'0');

    if (ele) 	dif(ele, st);

    return st;
}  

function kwUnscyncedTick(ele) {
    const lnow = new Date().getTime();
    const permi = 1000;
    let dint = permi - (lnow % permi);
    if (dint < 250) dint += 1000;

    setTimeout(function() { 
	kwsynclDisplayTick(lnow + dint, ele);
	kwUnscyncedTick(ele); 
    }, dint);
}

js/kwClockSyncedSafeInvoke.js
// execution begins at bottom
// I wrote a detailed readme file for doc.  I am Kwynn Buess of kwynn.com.

if (typeof byid !== 'function') { function byid(id) { return document.getElementById(id); } }

class kwclsySI { // Kwynn clock sync (or synced clock) safe invoke
constructor() {
    const testClassString = 'class kwclsySITest { doesPublicWork; }';
    const idjs = 'kwClockSynced.js';
    const thisSubPath = 'js/';
	
    try { 
		
	Function(testClassString)();
	
	const jsele = byid(idjs);
	const src = thisSubPath + idjs;
	jsele.src = src;

	jsele.onload = function() {
	    window.onload = function() {
		new kwclsyOntick(); 
	    }
	}   
	
	return;
	
    } catch (err) { console.info('synced clock will not work: ' + err.message); }

    // here we have fallen through to a non-synced clock
    window.onload = function() {  
	const idtcl  = 'kwTheClock';
	const ele = byid(idtcl);    
	if (!ele) {
	    console.info('no element ' + idtcl + ' exists, so nothing to do');
	    return;
	}

	if (ele && ele.dataset && ele.dataset.allow_unsynced != true) {
	    console.info('the allow_unsynced (clock) data / dataset attribute was not set');
	    return;
	}

	kwsynclDisplayTick(new Date().getTime(), ele);
	kwUnscyncedTick(ele);
    }
}
}

new kwclsySI(); // execution begins here


services/isAWS.php
<?php

// Kwynn 2020/01/25 9:10pm - goes to "is kwynn.com?" or at least a start
// echo Q | openssl s_client -showcerts -connect 127.0.0.1:443 2>/dev/null | grep 'CN = kwynn\.com'
// if use whole output, then "Verify return code: 0 (ok)" MIGHT be definitive

require_once('/opt/kwynn/kwutils.php');

class isAWSCmds { // see hu(man) message just below, and see
    // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
    
    const huMsg   = <<<CHMONE
  If the tests results are non-blank, these are indicators that the server is on Amazon Web Services.  This helps you further assess chrony accuracy.
CHMONE;
     
    const upfx    = 'http://169.254.169.254/latest/';
    const upfxmd  = self::upfx . 'meta-data/';
    const upfxcr  = self::upfx . '/dynamic/instance-identity/';
    const awspkf  = 'AWSPubKey_2020_01_1.txt';
    const razre   = '/^([a-z]{2}-[a-z]{4,30}-\d)[a-z]$/';
    const amire   = '/^(ami-)[0-9a-f]{8}/';
    const iddocnm = 'document';
    
    public static function gsc($cin, $pin = self::upfxmd) { return $pin . $cin;   }
    
    public static function gen($cmd, $regex, $tin = '') {

	try {
	    if ($tin) $txt = $tin;
	    else $txt = file_get_contents($cmd);
	    preg_match( $regex , $txt, $matches);
	    kwas(isset($matches[1]));
	    
	    if (self::testStatus() === 'showPrivate') $ri = 0;
	    else $ri = 1;
	    
	    return $matches[$ri];
	} catch (Exception $ex) { } return '';		
    }
        
    private static function doTests1() {
	
	if (!isAWS() && !self::testStatus()) return [];
	
	$iam = '  "InstanceProfileArn" : "arn:aws:iam::123456789012:instance-profile/abcd-ef"';

	$a = [
	    // ['/sys/hypervisor/uuid' , '/^(ec2)[a-f0-9-]{33}/', 'ec2223bc-f551-9e51-1d63-123456789abc'],
	    [isAWSCmds::gsc('iam/info'), '/  "InstanceProfileArn" : "(arn:aws:iam::)\d{12}/', $iam],
	    /* [isAWSCmds::gsc('ami-id')  , self::amire, 'ami-12345678'], // redundant, but I leave them for reference
	    [isAWSCmds::gsc('placement/availability-zone')  , self::razre, 'us-east-1x'] */
	];

	$r = [];


	foreach($a as $i) {
	    if (!isKwDev() || isAWS() || !self::testStatus()) $i[2] = '';
	    $t['cmd'] = '$ ' . $i[0];
	    $t['regex'] = $i[1];
	    $t['result'] = isAWSCmds::gen($i[0], $i[1], $i[2]);
	    $r[] = $t;
	}

	return $r;
}

public static function doTests() {
    
    $tn = 'isAWSCmds';
    $fs[] = [$tn, 'doTests1'];
    $fs[] = [$tn, 'crypto'];
    
    foreach($fs as $i=>$fv) $ra[] = $fs[$i]();
    
    $r2['isAWSTests'] = $ra;

    $r2['isAWSReadme'] = trim(self::huMsg);
    return $r2;    
}

private static function setTestA(&$a, $f, $re, $v) {
    if (!$re) return;
    kwas(preg_match($re, $v, $matches),"AWS $f field fails kwas 3");
    $r = [];
    $r['field']  = $f;
    $r['regex'] = $re;
    $r['result'] = 'passed regex';
    if (isset($matches[1])) $r['result'] = $matches[1];
    $a[] = $r;

    $x = 2;
}

public static function awsIDDoc() {
    
    if (!isAWS() && time() > strtotime('2020-01-17 03:00') && PHP_SAPI !== 'cli') return '';
    
    $lp = '/tmp/aid/awsIDDoc.txt';
    if (!isAWS() && file_exists($lp)) $txt = file_get_contents($lp); 
    else $txt = file_get_contents(self::upfxcr . self::iddocnm);
    
    $ra = json_decode($txt, 1);
    
    $iddts = [];
    
    $cks = ['accountId'		=> '/^\d{12}/',
	    'availabilityZone'	=> self::razre,
	    'imageId'           => self::amire,
	    'instanceId'        => '/^(i-)[a-z0-9]{8}/',
	    'pendingTime'       => '',
	    'instanceType'	=> '/^[a-z0-9]{2,10}\.[a-z0-9]{1,10}$/'];
    foreach($cks as $key => $re) {
	kwas(isset($ra[$key]), "AWS $key field fails kwas 1");
	$s = $ra[$key];
	kwas(is_string($s) && strlen($s) >= 5, "AWS $key field fails kwas 2");
	self::setTestA($iddts, $key, $re, $s);
	
    }
    
    $s = $ra['pendingTime'];
    $ts = strtotime($s);
    
    kwas($ts && $ts > 1106013719 /* 2005 */ && $ts < time() + 96000,'AWS pendingTime / time of last (re)start fails constraints');
    
    return ['txt' => $txt, 'rawarr' => $ra, 'publicRes' => $iddts];
}

public static function crypto() {
    
    exec('mktemp -d', $tda, $out); kwas($out === 0 && isset($tda[0]), 'mktemp failed');
    $td = $tda[0] . '/'; kwas(is_string($td) && trim(strlen($td)) > 12, 'mktemp string failed');
    
    $fs = [ 
	[__DIR__ . '/', self::awspkf],
	[self::upfxcr, 'pkcs7',  "-----BEGIN PKCS7-----\n", "\n-----END PKCS7-----\n"],
	// [self::upfxcr, 'document'] // redundant
    ];

    foreach($fs as $f) {
	if ($f[0] === self::upfxcr && !isAWS()) continue;
	$txt = file_get_contents($f[0] . $f[1]); kwas($txt, 'crypto file read failed');
	
	if (isset($f[2]) && isset($f[3])) {
	    $txt = $f[2] . $txt . $f[3];
	}
	
	$res = file_put_contents($td . $f[1], $txt); kwas($res, 'crypto file put failed');
	
    }
    
    $docr = self::awsIDDoc();
    
    if (isset($docr['txt'])) file_put_contents($td . self::iddocnm, $docr['txt']);
    
    
    // openssl smime -verify -in $PKCS7 -inform PEM -content $DOCUMENT -certfile AWSpubkey -noverify > /dev/null
    
    $c  = '';
    $c .= 'openssl smime -verify -in ';
    $c .= $td . $fs[1][1];
    $c .= ' -inform PEM -content ';
    $c .= $td . self::iddocnm;
    $c .= ' -certfile ';
    $c .= self::awspkf;
    $c .= ' -noverify ';
    $c .= ' 2>&1 1> /dev/null';
     
    $re = '/^(Verification successful)\s*$/';    
    $fr['cmd'] = $c;
    $fr['regex'] = $re;   
    $fr['result'] = '';
    
    $fr2 = [];
    
    if (isset($docr['publicRes'])) $fr2['fieldTests'] = $docr['publicRes'];
    $fr2['cryptoCmd'] = $fr;
    
    if (!isAWS())     return ['iddoc' => $fr2];

    $sres = shell_exec($c);
 
    preg_match($re, $sres, $matches);
    kwas(isset($matches[1]), 'AWS crypto match failed');

    $fr2['cryptoCmd']['result'] = $matches[1];
        
    return ['iddoc' => $fr2];
}

public static function testStatus() {
    
    global $argv;
    global $argc;
    
    if (PHP_SAPI !== 'cli') return false;
    if ($argc < 2 || $argv[1] !== 'test') return false;
    if ($argc > 2 && $argv[2] === 'showPrivate') return 'showPrivate';
    return 'test';    
}
}

if (isAWSCmds::testStatus()) {
    $r = isAWSCmds::doTests();
    var_dump($r);
    unset($r);
}
